### PR TITLE
feat: fast_exp

### LINKED ISSUE #700: bug: EXP will run out of resources on high exponent
# Bug Report

**Kakarot version:** [ee6458a](https://github.com/kkrt-labs/kakarot/commit/ee6458a7089f169334a7dba9665bcc5551507762)

**Current behavior:**
Given a high exponent value when calling the EXP opcode, the transaction will run out of resources. This is due to the fact that we don't use the fast powering algorithm for high exponent values. See code below:
https://github.com/kkrt-labs/kakarot/blob/95df46cf88a4225d26e7a642fe2d149b41c55249/src/kakarot/instructions/stop_and_arithmetic_operations.cairo#L462-L466

**Expected behavior:**
Exponentiation should work for high value of exponent.

**Steps to reproduce:**

- clone ef-tests: https://github.com/kkrt-labs/ef-tests
- generate katana dump of kakarot system and move the .katana folder to your ef-tests directory
- run 
```
TARGET=expPower256_d0g0v0_Shanghai cargo test -p ef-testing --features ef-tests -- --nocapture
```

Failing output should be:
```
2023-08-25T09:40:07.242339Z  WARN katana_core::backend::executor: Transaction execution error: "Error in the called contract (0x04e2460d0debdacdde372cb598b4e4403939ca112c8720e1c2aca13f6ac09b9a):
Error at pc=0:7:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:163)
Unknown location (pc=0:149)

Error in the called contract (0x04e2460d0debdacdde372cb598b4e4403939ca112c8720e1c2aca13f6ac09b9a):
Error at pc=0:31:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:4606)
Unknown location (pc=0:4545)
Unknown location (pc=0:4374)
Unknown location (pc=0:376)

Error in the called contract (0x049c085e55e5ab05ea46e16706720852421e86c1f8aeeab2be257313c4e290f4):
Error at pc=0:490:
Could not reach the end of the program. RunResources has no remaining steps.
Cairo traceback (most recent call last):
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18022)
Unknown location (pc=0:17250)
Unknown location (pc=0:9424)
"
[!] Case ~/code/rust/ef-tests/crates/ef-testing/ethereum-tests/BlockchainTests/GeneralStateTests/VMTests/vmArithmeticTest/expPower256.json failed (description: expPower256): Test failed: failed test expPower256_d0g0v0_Shanghai: expected storage value 0x0000000000000000000000000000000000000000000000000000000000000001, got 0x0000000000000000000000000000000000000000000000000000000000000000
```

### PR DESCRIPTION
**Pull request by Eikix**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #700

### CODE REVIEW

File: src/kakarot/instructions/stop_and_math_operations.cairo [hunk 1]
```diff
@@ -241,10 +241,11 @@ namespace StopAndMathOperations {
         EXP:
         let range_check_ptr = [ap - 2];
         let popped = cast([ap - 1], Uint256*);
+        tempvar bitwise_ptr = cast([fp - 4], BitwiseBuiltin*);
 
-        let result = uint256_exp(popped[0], popped[1]);
+        let result = uint256_fast_exp(popped[0], popped[1]);
```
Comment (Eikix) [inline src/kakarot/instructions/stop_and_math_operations.cairo:246]: alternatively we can do like SSJ ->
for popped[1] < 10, slow exp
for popped[1] >= 10, fast exp

but this adds one math cmp

File: src/utils/uint256.cairo [hunk 2]
```diff
@@ -1,21 +1,69 @@
...
+    if (exponent_is_zero != FALSE) {
+        return one;
+    }
+
+    let (exponent_is_one) = uint256_eq(exponent, one);
+    if (exponent_is_one != FALSE) {
+        return value;
+    }
+
+    let (exponent_is_odd) = uint256_and(exponent, one);
```
Comment (ClementWalter) [inline src/utils/uint256.cairo:55]: I think that div_rem is cheaper than bitwise, but it changed, not sure about the diff now

but we wouldn't need the bitwise_ptr op at all, so save again a bit

File: src/utils/uint256.cairo [hunk 3]
```diff
@@ -1,21 +1,69 @@
...
+    if (exponent_is_odd.low != FALSE) {
```
Comment (ClementWalter) [inline src/utils/uint256.cairo:56]: why do you first check parity while it seems like you could do it afterwards using the rem?

File: src/utils/uint256.cairo [hunk 4]
```diff
@@ -1,21 +1,67 @@
-from starkware.cairo.common.uint256 import Uint256, uint256_eq, uint256_le, uint256_sub, uint256_mul
+from starkware.cairo.common.uint256 import (
+    Uint256,
+    uint256_eq,
+    uint256_le,
+    uint256_sub,
+    uint256_mul,
+    uint256_unsigned_div_rem,
+    uint256_and,
```
Comment (ClementWalter) [inline src/utils/uint256.cairo:8]: remove unused imports

File: src/utils/uint256.cairo [hunk 5]
```diff
@@ -1,21 +1,67 @@
...
+    let (res, _) = uint256_mul(value, pow);
+    return res;
+}
+
+// @notice Internal fast exponentiation of two 256-bit integers.
+// @dev The result is modulo 2^256.
+// @param value - The base.
+// @param exponent - The exponent.
+// @return The result of the exponentiation.
+func uint256_fast_exp{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(
```
Comment (ClementWalter) [inline src/utils/uint256.cairo:37]: remove unused implitic

**Review by ClementWalter — DISMISSED** [review]

**Review by ClementWalter — APPROVED** [review]

**Comment by greged93** [comment]
I think we could try to uncomment some of the exponentiation tests from the ef-tests with this

**Pull request merged**